Add RenameModule command

Kaudaj opened the pull request:
This command rename a module by :

- Uninstalling the module
- Replacing all occurences in its files
- Run `composer update` and `composer dumpautoload -a`
- Installing the module

You can change the author name with `new-author` option and add any extra replacement pairs with `extra-replacements` option.
You can also choose to keep the old module installed with `keep-old` option.

SebSept commented:
I'll check on Monday evening.
Thanks for contribution :)

Kaudaj commented:
> 
> 
> I'll check on Monday evening.
> Thanks for contribution :)

No problem!
I don't really understand tests failure by the way, since I did composer update on my side.

okom3pom commented:
Hello, 

Tests are running with php 7.2 

https://github.com/friends-of-presta/fop_console/blob/dev/.github/workflows/phpstan.yml#L35-L38

https://github.com/friends-of-presta/fop_console/blob/dev/.github/workflows/php-cs-fixer.yml#L19-L22

https://github.com/friends-of-presta/fop_console/blob/dev/.github/workflows/lint_yaml.yml#L19-L22

The new composer.lock need php 7.3 with : 

https://github.com/Kaudaj/fop_console/blob/rename-module/composer.lock#L92-L94

I think we can use php 7.3 for testing. But we have to keep in mind that we support prestashop 1.7.5 which uses php 7.2

okom3pom reviewed (changes requested):
This command is very useful and nice.
I tested the PR, it works well with modules without prefixes but with prefixes it doesn't work.

If I add the _ to the prefix it works but only with the same prefix, It is the prestashop naming that is the problem `Ps_Checkpayment`

I running test with php 7.3 it's work fine : https://github.com/okom3pom/fop_console/runs/3039209619

Kaudaj commented:
> 
> 
> Hello,
> 
> Tests are running with php 7.2
> 
> https://github.com/friends-of-presta/fop_console/blob/dev/.github/workflows/phpstan.yml#L35-L38
> 
> https://github.com/friends-of-presta/fop_console/blob/dev/.github/workflows/php-cs-fixer.yml#L19-L22
> 
> https://github.com/friends-of-presta/fop_console/blob/dev/.github/workflows/lint_yaml.yml#L19-L22
> 
> The new composer.lock need php 7.3 with :
> 
> https://github.com/Kaudaj/fop_console/blob/rename-module/composer.lock#L92-L94
> 
> I think we can use php 7.3 for testing because php 7.2 is EOL. But we have to keep in mind that we support prestashop 1.7.5 which uses php 7.2

Ups! Alright, I reverted composer.lock and added composer/composer dependency without updating all packages.

Kaudaj commented on src/Commands/Modules/RenameModule.php:
```diff
@@ -0,0 +1,322 @@
+        if ($newModule && $newModule->install()) {
+            $io->success('The fresh module ' . $newModuleName . ' has been installed.');
+        } else {
+            $io->error('The module ' . $oldModuleName . ' couldn\'t be installed.');
```
Fixed it so old modules containing underscore in their name can be renamed too.
Thanks for pointing it out!

Kaudaj commented on src/Commands/Modules/RenameModule.php:
```diff
@@ -0,0 +1,322 @@
+            return false;
+        }
+
+        $prefix = count($fullName) == 2 ? $fullName[0] : '';
```
Since I try to respect the latest PrestaShop conventions as much as possible (and I would like to encourage people to do so), I have decided to comply with the requirements of the new translation system:

> Only word characters (A-z, 0-9) are supported.
> Make sure that the technical name of your module does not contain contains underscores (_) or any other unsupported symbol, or else translation may not work.

Should I add an `underscore` option to make add one in the new module name?

okom3pom replied:
No you are right :+1:  but in the arguments i see : 

`Module new name with following format : Prefix_ModuleNewNameCamelCased`

I just tested again with : `../../bin/console fop:modules:rename ps_emailsubscription okommail --keep-old`

`Ps_Emailsubscription` is not proposed for the  occurence i think this bug is only with prestashop modules.

![Capture d’écran de 2021-07-11 16-57-39](https://user-images.githubusercontent.com/5064590/125200015-3a26d900-e269-11eb-8b2f-e8053b9164f3.png)

Kaudaj replied:
Ah no, this is not a bug!
You have to camel case the names, otherwise the command won't know the different "words" of the module. For this one, it's not so important but with Ps_CustomerSignIn for example, it is.
So with ps_emailsubscription, here is the right command (assuming you want okom as prefix and mail as name for the new module) :

`../../bin/console fop:modules:rename Ps_Emailsubscription okom_mail --keep-old`

I get the following result:

![image](https://user-images.githubusercontent.com/48719560/125201856-3bf49a80-e271-11eb-9373-54e2c8b6cb70.png)

The fresh module is installed on my side, it should work if you camel case the module names.

okom3pom replied:
Okay now I understand how it works :+1: 

Whyt not for the old module to say use the class name because for me module name is `$this->name = 'ps_emailsubscription';`

It's work good very nice job.

Kaudaj replied:
Well, I thought about it too but if the old module name has no underscore in its full name, how can we know where the prefix ends? Here are some examples, where the prefix is "Fop" and the name is "ModuleExample" :

- FopModuleExample
- FOPModuleExample
- FoPModuleExample

I could assume that everyone uses the same format in their module prefix, but I prefer to let the user choose their format freely. I added warnings if names are not camel cased (no upper case letter in the full name).

okom3pom commented:
One last feedback, you need to change the Readme because the commands change.

Kaudaj commented:
> 
> 
> One last feedback, you need to change the Readme because the commands change.

Done!
I'll try to do an other PR to rename every command with corresponding domain when I have some time by the way.

okom3pom commented:
Rebase this day a PR update the readme :)

I already used your command 2 times this morning :+1:

SebSept commented:
The 2 first commits are not related to your command directly.
It's better to make a separed PR, Single Responsibility Principle 😉 
But that's ok for this time, I don't want to scare you 😱 😉

SebSept commented on src/Commands/Modules/RenameModule.php:
```diff
@@ -0,0 +1,376 @@
+            exec('rm -rf vendor');
+            $installCommand = new ArrayInput(['command' => 'update']);
+            $dumpautoloadCommand = new ArrayInput(['command' => 'dumpautoload', '-a']);
+            $application = new Application();
```
The less dependencies there is the better it is (less problem).
Did you tried with `Symfony\Component\Process\Process` ? (I've used it some times and it works fine).

Kaudaj commented:
> 
> 
> The 2 first commits are not related to your command directly.
> It's better to make a separed PR, Single Responsibility Principle 😉
> But that's ok for this time, I don't want to scare you 😱 😉

Ah yep, my bad sorry ^^'
I'll do the PR on the [right naming conventions](https://github.com/friends-of-presta/fop_console/issues/85) anyway!

okom3pom commented:
ping @SebSept @jf-viguier @nenes25 

This command is very useful.  Plz approve :)

nenes25 commented:
I've tried and the module works fine ! Good Job !

Regarding the code, i think that there are too much lines in the function execute.
The code could be split in small and more readable sub functions.
Moreover i'm not sure that deleting functions using " rm -rf " are working on windows environment ( maybe @jf-viguier can confirm ? )

Kaudaj commented:
@nenes25 Done! 

- [Refactor command execution in subfuctions](https://github.com/friends-of-presta/fop_console/pull/154/commits/228a4ac643ef8b01ee7a5c51a18c7892186f467d)*
- [Provide Windows support](https://github.com/friends-of-presta/fop_console/pull/154/commits/89bc8dd659e7b90d0818146a014d466935e001d9)

I would like to have your opinion on:

- the $input and $output argument in each function. Would it be better to make it class member variables?
- the $moduleInfos member variables. It looks not very safe and easy to maintain. Should I make a class? Should I keep this ambiguous name (can't find a better one atm)?

Kaudaj commented on src/Commands/Modules/RenameModule.php:
```diff
@@ -0,0 +1,376 @@
+            exec('rm -rf vendor');
+            $installCommand = new ArrayInput(['command' => 'update']);
+            $dumpautoloadCommand = new ArrayInput(['command' => 'dumpautoload', '-a']);
+            $application = new Application();
```
Done! : [Replace composer dependency by Process usage ](https://github.com/friends-of-presta/fop_console/pull/154/commits/89bc8dd659e7b90d0818146a014d466935e001d9)

Kaudaj commented on src/Commands/Modules/RenameModule.php:
```diff
@@ -0,0 +1,376 @@
+                InputArgument::REQUIRED,
+                'Module new name with following format : Prefix_ModuleNewNameCamelCased'
+            )
+            ->addUsage('--new-author=[AuthorCamelCased]')
```
Done! : [Improve command configuration readability ](https://github.com/friends-of-presta/fop_console/pull/154/commits/d89a65f0817cb44126a3ea119800ffda30792fe5

Kaudaj commented on src/Commands/Modules/RenameModule.php:
```diff
@@ -0,0 +1,376 @@
+                InputArgument::REQUIRED,
+                'Module new name with following format : Prefix_ModuleNewNameCamelCased'
+            )
+            ->addUsage('--new-author=[AuthorCamelCased]')
```
Done! : https://github.com/friends-of-presta/fop_console/pull/154/commits/d89a65f0817cb44126a3ea119800ffda30792fe5

Kaudaj commented on src/Commands/Modules/RenameModule.php:
```diff
@@ -0,0 +1,376 @@
+        $io = new SymfonyStyle($input, $output);
+
+        if (!($oldFullName = $this->mapModuleName($input->getArgument('old-name')))) {
+            $io->error('Please check the required format for the module old name');
```
Looks outdated: I added concrete usage examples in '--help' text in the meantime and more explicit errors in https://github.com/friends-of-presta/fop_console/pull/154/commits/cf4adcd19457f37c524c927bf9dcea90d326f33e.

Kaudaj commented on src/Commands/Modules/RenameModule.php:
```diff
@@ -0,0 +1,376 @@
+            ->addArgument(
+                'old-name',
+                InputArgument::REQUIRED,
+                'Module current name with following format : Prefix_ModuleCurrentNameCamelCased'
```
Yes precisely, you have to use the module class name instead of the module name.
I tried to make it clearer in https://github.com/friends-of-presta/fop_console/pull/154/commits/46afc4b8d5df0e4cc0b06a869f24758f3f9cbc5c, even if it's still not ideal. If you have suggestions to make it even clearer, please let me know.

Kaudaj commented on src/Commands/Modules/RenameModule.php:
```diff
@@ -0,0 +1,376 @@
+    {
+        $io = new SymfonyStyle($input, $output);
+
+        if (!($oldFullName = $this->mapModuleName($input->getArgument('old-name')))) {
```
Fixed in https://github.com/friends-of-presta/fop_console/pull/154/commits/cf4adcd19457f37c524c927bf9dcea90d326f33e!

Kaudaj commented on src/Commands/Modules/RenameModule.php:
```diff
@@ -0,0 +1,376 @@
+        return 0;
+    }
+
+    public function mapModuleName($arg)
```
Fixed in https://github.com/friends-of-presta/fop_console/pull/154/commits/cf4adcd19457f37c524c927bf9dcea90d326f33e!

SebSept commented:
nice work 👍🏼
